Fix adding include flags

anmonteiro opened the pull request:
Not really sure what to call this PR.

Fixes a regression introduced by https://github.com/ocaml/dune/pull/2920 and released in 2.1.0.

I'm opening this PR to get some early feedback while I try to find a minimal reproduction.

The failure mode for me is Dune failing to link an executable that depends on `libbigstringaf_stubs`.

anmonteiro commented:
I just realized that this is what's blocking Dune 2.1.0 to get released to OPAM: https://github.com/ocaml/opam-repository/pull/15590#issuecomment-568247747

anmonteiro commented:
This can easily be reproduced with the following dune file and an empty `test.ml` file:

```lisp
(executable
 (name test)
 (libraries bigstringaf))
```

but I have only managed to repro this with a dependency external to the project (i.e. if I define a library with stubs inside the same dune workspace it links correctly).

I'm out of ideas for writing a test, so any guidance would be appreciated now.

rgrinberg commented:
We have some tests where dune builds a library and then it is imported to another workspace as external. See `variants/run.t` for some examples

anmonteiro committed:
Add test skeleton

Signed-off-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>

anmonteiro commented:
I added a test like you suggested, but I still fail to reproduce the issue with it. Is there anything that I missed in the test?

rgrinberg commented:
> I added a test like you suggested, but I still fail to reproduce the issue with it. Is there anything that I missed in the test?

Isn't it because you're testing for the bug after you've already fixed it?

Could you sign your commits please?

anmonteiro commented:
> Isn't it because you're testing for the bug after you've already fixed it?

If you cherry pick that last commit on master it passes.

> Could you sign your commits please?

rebasing now

rgrinberg commented:
I've tried to modify the tests to reflect the setup with bigstringaf but without success. Do you have an environment where I can reproduce your issue?

anmonteiro commented:
if you `opam install bigstringaf` and apply the following patch, the `@link-includes` test fails without my patch:

```diff
diff --git a/test/blackbox-tests/test-cases/link-includes/run.t b/test/blackbox-tests/test-cases/link-includes/run.t
index 39dda780..f7c9f472 100644
--- a/test/blackbox-tests/test-cases/link-includes/run.t
+++ b/test/blackbox-tests/test-cases/link-includes/run.t
@@ -39,14 +39,14 @@ First we create an external library and implementation
   $ mkdir exe
   $ echo "(lang dune 2.0)" > exe/dune-project
   $ cat >exe/dune <<EOF
-  > (executable (name bar) (libraries dunetestlib2))
+  > (executable (name bar) (libraries bigstringaf dunetestlib2))
   > EOF
   $ cat >exe/bar.ml <<EOF
   > let () = Lib2.run ()
   > EOF
 
 Then we make sure that it works fine.
-  $ env OCAMLPATH=lib1/_build/install/default/lib:lib2/_build/install/default/lib dune exec --root exe ./bar.exe
+  $ env OCAMLPATH=~/.opam/4.09.0/lib/bigstringaf:lib1/_build/install/default/lib:lib2/_build/install/default/lib dune exec --root exe ./bar.exe
   Entering directory 'exe'
   Entering directory 'exe'
   42
```

rgrinberg commented:
Okay, I reproduced the bug. The issue is with `# JBUILDER_GEN`

@snowleopard could you review?

anmonteiro commented:
Thanks for taking the time to investigate this further. I'm not sure how those pieces interact, but I defer to more knowledgeable folks.

rgrinberg committed:
Better reading of findlib stubs

When dune reads libraries using META files, it ignores all stub
archives, and native archives. Instead, we can make dune a bit smarter
and try to guess this information.

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg commented:
I reworked the fix to be less invasive. Let's add @diml as a reviewer as well. Guessing artifacts for META defined libraries is a bit brittle, but it fixes the problem in more contexts and doesn't just add (sometimes) unnecessary flags.

snowleopard commented:
@rgrinberg @anmonteiro Many thanks for doing the detective work and for the fix! Could you please explain the issue and the solution in more detail?

Let's take this `dune` file as an example:

```
(executable
 (name test)
 (libraries bigstringaf))
```

I'd like to understand:

* How did #2920 break it? Did it remove the `-Lpath/to/libbigstringaf_stubs` flag? Where did this flag come from before #2920?
* How does the proposed solution work? As far as I understand the key is [this guessing bit](https://github.com/ocaml/dune/pull/2994/files#diff-b53c2884db88e2d42607e264426b3f23R286-R310) but it looks a bit mysterious. It would be great to explain how it works in a comment.

rgrinberg commented:
> How did #2920 break it? Did it remove the -Lpath/to/libbigstringaf_stubs flag? Where did this flag come from before #2920?

I've commented on the line that removed the `c_include_path` from the link flags in that PR.

> How does the proposed solution work? As far as I understand the key is this guessing bit but it looks a bit mysterious. It would be great to explain how it works in a comment.

The idea is: archives, object files, and shared objects all have recognizable extensions. For libraries that do not document their build artifacts properly (those using META templates or non dune libraries), we attempt to reconstruct these by guessing what those artifacts would be using extensions. This has a syscall cost, but it's not so bad since the number of such libraries is tiny by now.

rgrinberg commented:
I've reworked the discovery of archives to only work on stub and native archives. Everything else was wrong and/or unnecessary to fix this bug.

snowleopard commented on src/dune/findlib/findlib.ml:
```diff
@@ -282,14 +283,40 @@ module Package = struct
+      let foreign_archives, native_archives =
+        match Path.readdir_unsorted t.dir with
+        | Error _ ->
+          (* We could add a warning like this: User_warning.emit
```
Shouldn't this actually be an error? Can a `Package` have a non-existing/non-readable `dir`?

rgrinberg replied:
This was the initial behavior for due but it turned out that it's not what we want in many cases. Sometimes, there are some permission errors and the library in question isn't used to build anything anyway. We ended up deciding to delay all errors until build time. In this case, we'll get an error anyway when we try to read any artifact of this library eventually.

snowleopard replied:
Aha, I see! I think your answer should go into the comment for the `Error _` case. This is a useful piece of information but it will get lost and forgotten very quickly.

rgrinberg replied:
Done

snowleopard commented on src/dune/findlib/findlib.ml:
```diff
@@ -282,14 +283,40 @@ module Package = struct
       let known_implementations = P.Map.empty in
       let default_implementation = None in
       let wrapped = None in
+      let foreign_archives, native_archives =
```
```suggestion
      (* Here we scan [t.dir] and consider all files named [lib*.ext_lib] to be
         foreign archives, and all other files with the extension [ext_lib] to
         be native archives. The resulting lists of archives will be used to
         compute appropriate flags for linking dependant executables. *)
      let foreign_archives, native_archives =
```

rgrinberg replied:
Added comment manually

snowleopard replied:
Thanks!

snowleopard commented:
> I've commented on the line that removed the `c_include_path` from the link flags in that PR.

Aha, thanks! Am I right in understanding that using C include path information for linking was a bit of a hack but it did the job for libraries that do not document their build artifacts properly?

> I've reworked the discovery of archives to only work on stub and native archives. Everything else was wrong and/or unnecessary to fix this bug.

Just to clarify: is extending this mechanism to also handle `dll*.ext_dll` files wrong or unnecessary?

P.S.: I've added a comment via GitHub suggestions -- it will certainly help the future me to understand what's going on.

rgrinberg commented:
> Aha, thanks! Am I right in understanding that using C include path information for linking was a bit of a hack but it did the job for libraries that do not document their build artifacts properly?

That's right.

> Just to clarify: is extending this mechanism to also handle dll*.ext_dll files wrong or unnecessary?

It seems like we don't read the stub dll even for proper dune defined external libraries. So we are no worse than before for findlib libraries.

snowleopard commented on CHANGES.md:
```diff
@@ -21,6 +21,9 @@
   dependencies so that we can have things such as `(<> :os win32)`
   (#2965, @diml)
 
+- Guess foreign archives, native archives, foreign dll files for libraries read
```
We no longer guess foreign dll files.

snowleopard reviewed (approved):
Looks good now, thank you! I'd like @diml to look at the PR too in case I missed some corner case.

diml commented:
Just to make sure I understand what's going on:

- before #2920, we were systematically passing `-I <dir>` for each library dependency at link time
- since #2920, we only pass `-I <dir>` for library dependency that have we know C stubs

For external libraries that were not built with Dune, we don't know whether they have C stubs or not and as a result we were not passing `-I <dir>`. This is a problem for non-dune external libraries that have C stubs.

However, in the opam-repository PR, the problem is that we are not passing `-I <dir>` for bin_prot which is built with Dune. Do we know why that is?

The guessing code seems like it could be wrong in some cases, for instance if the library is called `libfoo` then `libfoo.a` will be considered as a foreign archives while it is instead the native one. At the same time, this only affects non-dune libraries so it's not the end of the world either. We could also mark such libraries as "might have foreign archives".

rgrinberg commented:
> However, in the opam-repository PR, the problem is that we are not passing -I <dir> for bin_prot which is built with Dune. Do we know why that is?
Because it has a META template. Therefore it’s installed as a findlib library
On Jan 6, 2020, 10:41 PM +0700, Jérémie Dimino <notifications@github.com>, wrote:
> Just to make sure I understand what's going on:
>
> • before #2920, we were systematically passing -I <dir> for each library dependency at link time
> • since #2920, we only pass -I <dir> for library dependency that have we know C stubs
>
> For external libraries that were not built with Dune, we don't know whether they have C stubs or not and as a result we were not passing -I <dir>. This is a problem for non-dune external libraries that have C stubs.
> However, in the opam-repository PR, the problem is that we are not passing -I <dir> for bin_prot which is built with Dune. Do we know why that is?
> The guessing code seems like it could be wrong in some cases, for instance if the library is called libfoo then libfoo.a will be considered as a foreign archives while it is instead the native one. At the same time, this only affects non-dune libraries so it's not the end of the world either. We could also mark such libraries as "might have foreign archives".
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

diml commented:
Make sense. No objection on my side them.

anmonteiro commented:
Thanks for your help on this one, @rgrinberg!
